### PR TITLE
Serial refactor

### PR DESCRIPTION
**Pull request by zfields**
@ras0219-msft @turkycat @apdutta

### CODE REVIEW

File: source/Serial/BluetoothSerial.cpp [hunk 1]
```diff
@@ -87,293 +86,218 @@ BluetoothSerial::BluetoothSerial(
...
+    // Ensure known good state
+    end();
+
+    // Although this path is not optimal, the behavior (the calls must be made from the UI thread) is mandated by the bluetooth API. The algorithm is a compromise to provide the succint, maintainable code.
+    Concurrency::create_task(listAvailableDevicesAsync())
+        .then([this](Windows::Devices::Enumeration::DeviceInformationCollection ^device_collection_)
+    {
+        // If a friendly name was specified, then identify the associated device
+        if (_device_name) {
+            // Store parameter as a member to ensure the duration of object allocation
```
Comment (ras0219-msft) [inline source/Serial/BluetoothSerial.cpp]: Might throw a debug assert that ```_device == nullptr``` in this case

File: source/Serial/BluetoothSerial.cpp [hunk 2]
```diff
@@ -87,293 +86,218 @@ BluetoothSerial::BluetoothSerial(
...
-	{
-		if( device->Id->Equals( _deviceIdentifier->ToString() ) || device->Name->Equals( _deviceIdentifier ) )
-		{
-			return device;
-		}
-	}
-	return nullptr;
-}
\ No newline at end of file
+    for (auto &&device : devices_)
```
Comment (ras0219-msft) [inline source/Serial/BluetoothSerial.cpp]: Good to know this works on WinRT collections.

File: source/Serial/BluetoothSerial.cpp [hunk 3]
```diff
@@ -87,293 +86,218 @@ BluetoothSerial::BluetoothSerial(
...
-	UNREFERENCED_PARAMETER( config_ );
-	begin( true );
+    // Discard incoming parameters inherited from IStream interface.
+    UNREFERENCED_PARAMETER(baud_);
+    UNREFERENCED_PARAMETER(config_);
+
+    // Ensure known good state
+    end();
+
+    // Although this path is not optimal, the behavior (the calls must be made from the UI thread) is mandated by the bluetooth API. The algorithm is a compromise to provide the succint, maintainable code.
```
Comment (connerbrooks) [inline source/Serial/BluetoothSerial.cpp]: 80 char line limit police :cop: :oncoming_police_car:

File: Microsoft.Maker.win10/Microsoft.Maker.Serial/Microsoft.Maker.Serial.vcxproj.filters [hunk 4]
```diff
@@ -10,11 +10,13 @@
     <ClCompile Include="pch.cpp" />
     <ClCompile Include="..\..\source\Serial\BluetoothSerial.cpp" />
     <ClCompile Include="..\..\source\Serial\USBSerial.cpp" />
+    <ClCompile Include="..\..\source\Serial\DfRobotBleSerial.cpp" />
```
Comment (ooeygui) [inline Microsoft.Maker.win10/Microsoft.Maker.Serial/Microsoft.Maker.Serial.vcxproj.filters]: Do we have permission to use this name in our sources?

File: source/Serial/DfRobotBleSerial.cpp [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+uint32_t
+DfRobotBleSerial::write(
+    uint8_t c_
+    )
+{
+    // Check to see if connection is ready
+    if (!connectionReady()) { return 0; }
+
+    _tx->WriteByte(c_);
+    create_task(_gatt_characteristic->WriteValueAsync(_tx->DetachBuffer(), GattWriteOption::WriteWithResponse))
```
Comment (connerbrooks) [inline source/Serial/DfRobotBleSerial.cpp]: WriteWithResponse sends data out of order, using WriteWithoutResponse sends the data in order.

**Comment by turkycat** [comment]
Please rebase your commits on top of \ms-iot\.

**Comment by connerbrooks** [comment]
Previous comment is incorrect, everything is sent out of order, we need to create some sort of Virtual Serial interface that forces byte order.

File: source/Serial/BluetoothSerial.cpp (hunk 2 above)
Comment (zfields) [inline source/Serial/BluetoothSerial.cpp]: :+1:

File: source/Serial/BluetoothSerial.cpp (hunk 3 above)
Comment (zfields) [inline source/Serial/BluetoothSerial.cpp]: so vim

File: Microsoft.Maker.win10/Microsoft.Maker.Serial/Microsoft.Maker.Serial.vcxproj.filters (hunk 4 above)
Comment (zfields) [inline Microsoft.Maker.win10/Microsoft.Maker.Serial/Microsoft.Maker.Serial.vcxproj.filters]: Good call! We did talk with him face-to-face about what we were planning do to, but we didn't cover specifics. I'll make sure and have Paul ask him tomorrow, before we pull this in.

File: source/Serial/DfRobotBleSerial.cpp (hunk 5 above)
Comment (zfields) [inline source/Serial/DfRobotBleSerial.cpp]: We have discovered this is a batching error, so I'm separating the flush() operation from the write.

**Comment by zfields** [comment]
@turkycat knitpick1 - I favor variable names in alphabetical order, not aesthetic. When you are debugging, you're already having a hard time, and the variables should be easy to lookup.

**Comment by zfields** [comment]
@turkycat knitpick2 - Hate tabs, love spaces. If there are tabs mixed in, then I wasn't able to find them.

File: source/Serial/BluetoothSerial.cpp (hunk 1 above)
Comment (zfields) [inline source/Serial/BluetoothSerial.cpp]: What's the best syntax for this?

File: Microsoft.Maker.win10/Microsoft.Maker.Serial/Microsoft.Maker.Serial.vcxproj.filters (hunk 4 above)
Comment (zfields) [inline Microsoft.Maker.win10/Microsoft.Maker.Serial/Microsoft.Maker.Serial.vcxproj.filters]: I spoke with the DF Robot crew. The "Serial Port" service used is defined by DF Robot and our file is specific to their implementation, so they are happy with the naming convention.

**Pull request merged**